### ISSUE TITLE
Error to compile latest version of APSIMX

### ISSUE THREAD

**Issue by byzheng** [APSIMInitiative/ApsimX#3532]
The new clone of APSIMX and APSIM.Shared cannot be complied after returning from two months.

It seems some packages are missing, but I can find them in the package folders.

Any ideas about these errors?

The errors list below.
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0246	The type or namespace name 'DocumentFormat' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	17	Active
Error	CS0246	The type or namespace name 'DocumentFormat' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	18	Active
Error	CS0246	The type or namespace name 'ExcelDataReader' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	20	Active
Error	CS0246	The type or namespace name 'FirebirdSql' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	13	Active
Error	CS0246	The type or namespace name 'FbConnection' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	25	Active
Error	CS0246	The type or namespace name 'FbServerType' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	30	Active
Error	CS0246	The type or namespace name 'FbTransaction' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	322	Active
Error	CS0246	The type or namespace name 'SpreadsheetDocument' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	46	Active
Error	CS0103	The name 'SpreadsheetDocument' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	46	Active
Error	CS0246	The type or namespace name 'Sheets' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	48	Active
Error	CS0246	The type or namespace name 'Sheet' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	49	Active
Error	CS1579	foreach statement cannot operate on variables of type 'Sheets' because 'Sheets' does not contain a public instance definition for 'GetEnumerator'	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	49	Active
Error	CS0246	The type or namespace name 'IExcelDataReader' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	78	Active
Error	CS0103	The name 'ExcelReaderFactory' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\ExcelUtilities.cs	78	Active
Error	CS0246	The type or namespace name 'FbConnection' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	25	Active
Error	CS0103	The name 'FbServerType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	30	Active
Error	CS0103	The name 'FbConnection' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	52	Active
Error	CS0246	The type or namespace name 'FbConnectionStringBuilder' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	71	Active
Error	CS0246	The type or namespace name 'FbConnectionStringBuilder' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	71	Active
Error	CS0246	The type or namespace name 'FbCommand' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	144	Active
Error	CS0246	The type or namespace name 'FbCommand' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	144	Active
Error	CS0246	The type or namespace name 'FbCommand' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	267	Active
Error	CS0246	The type or namespace name 'FbCommand' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	267	Active
Error	CS0246	The type or namespace name 'FbDataAdapter' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	273	Active
Error	CS0246	The type or namespace name 'FbDataAdapter' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	273	Active
Error	CS0246	The type or namespace name 'FbCommand' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	324	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	333	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	339	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	344	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	349	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	354	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	359	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	364	Active
Error	CS0103	The name 'FbDbType' does not exist in the current context	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	367	Active
Error	CS0246	The type or namespace name 'FbTransaction' could not be found (are you missing a using directive or an assembly reference?)	APSIM.Shared	C:\Users\zhe00a\Documents\Working\04-Software\032-ApsimX\APSIM.Shared\Utilities\Firebird.cs	525	Active

```

**Comment by byzheng** [comment]
Seems I have to copy all packages from ApsimX\packages into APSIM.Shared. It is strange as I don't think anything is upgraded since latest pull. 

The version information of my Visual Studio.
```
Microsoft Visual Studio Community 2017 
Version 15.7.3
VisualStudio.15.Release/15.7.3+27703.2026
Microsoft .NET Framework
Version 4.7.02556

Installed Version: Community

ASP.NET and Web Tools 2017   15.0.40522.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.7.31476
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.60419.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.0.40424.0
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   15.0.40424.0
Azure Functions and Web Jobs Tools

C# Tools   2.8.3-beta6-62923-07. Commit Hash: 7aafab561e449da50712e16c9e81742b8e7a2969
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   15.7.18131.1
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

JavaScript UWP Project System   2.0
JavaScript UWP Project System

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.10420.2

Microsoft Continuous Delivery Tools for Visual Studio   0.3
Simplifying the configuration of continuous build integration and continuous build delivery from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

MySQL for Visual Studio   1.2.8
Data design and management tools for MySQL. Copyright (c) 2007, 2018, Oracle and/or its affiliates. All rights reserved.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Python   15.7.18131.1
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Python - Django support   15.7.18131.1
Provides templates and integration for the Django web framework.

Python - IronPython support   15.7.18131.1
Provides templates and integration for IronPython-based projects.

Python - Profiling support   15.7.18131.1
Profiling support for Python projects.

R Tools for Visual Studio   1.3.40517.1016
Provides project system, R Interactive window, plotting, and more for the R programming language.

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   15.1.61804.210
Microsoft SQL Server Data Tools

TypeScript Tools   15.7.20419.2003
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.8.3-beta6-62923-07. Commit Hash: 7aafab561e449da50712e16c9e81742b8e7a2969
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.1 for F# 4.1   15.7.0.0.  Commit Hash: 2527e6829ecdc8281ee60d83be8cfd0fa720a648.
Microsoft Visual F# Tools 10.1 for F# 4.1

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for Universal Windows Apps   15.0.27703.2026
The Visual Studio Tools for Universal Windows apps allow you to build a single universal app experience that can reach every device running Windows 10: phone, tablet, PC, and more. It includes the Microsoft Windows 10 Software Development Kit.

Windows Machine Learning Generator Extension   1.0
Windows Machine Learning Visual Studio Extension Detailed Info
```

**Comment by hol430** [comment]
I'm getting this as well on a fresh clone. As a workaround until I fix this, you should be able to open APSIM.Shared.sln in visual studio, click tools -> nuget package manager -> manage nuget packages for solution -> restore nuget packages.

**Comment by IrisKydH** [comment]
I've encountered the same issue. Have tried both solutions abovementioned to fix it, but it seems not to work for me. Any other ideas?

**Comment by byzheng** [comment]
@IrisKydH Try to compile the APSIM.shared firstly?

**Comment by hol430** [comment]
Does the packages directory exist at APSIM.Shared/packages? If it does exist, you must be running into a different issue. Also, which operating system are you using? I've only tested my solution on Windows.

**Comment by IrisKydH** [comment]
@byzheng Thanks for your advice, compiled the APSIM.shared firstly and then run into some other issues, trying to fix them.

**Comment by IrisKydH** [comment]
@hol430 I'm using windows and Visual Studio 2017, I encounter some new issues, try to figure out why.

**Comment by hol430** [comment]
What issues are you running into? Can you paste an error message?

**Comment by byzheng** [comment]
@hol430 yes the packages directory existd at APSIM.Shared/packages. I guess some update of VS or windows changes the search path of packages folders.

**Comment by hol430** [comment]
Yeah originally APSIM.Shared used to go looking inside the ApsimX/packages for its nuget packages. However, this has 2 drawbacks:

1. Can't compile APSIM.Shared unless ApsimX sits next to it on disk
2. Any nuget packages used in APSIM.Shared must also be used in ApsimX

I've recently modified the APSIM.Shared project file so that it looks inside APSIM.Shared/packages for its nuget packages. This has turned out to not be a bad idea though, as msbuild/nuget don't seem to automatically restore nuget packages for APSIM.Shared when we build it as part of ApsimX. I'm guessing this is because it's in a separate directory, but I'm not really sure. 

As I said a few comments above, doing a manual nuget restore (via the command line tool, or by opening APSIM.Shared.sln in visual studio) fixes the problem for me but it's pretty awkward. 

I'm not really sure what the best solution is... one solution would be to add a new pre-build step to APSIM.Shared.csproj which restores the nuget packages. This seems to work for me, but I'm guessing most people don't have the nuget command line tool installed and on PATH. 

For now I guess the simplest solution would be to revert my recent change and let APSIM.Shared continue to look inside ApsimX/packages for its nuget packages.

**Comment by IrisKydH** [comment]
@byzheng @hol430 It finally works! Thank you so much.
@zur003 Yea, make sense, I'm using the version 15.9.7.

**Comment by BenAbabaei** [comment]
I rebuilt APSIM.Shared and it worked for me. I noticed that now **Models** is the startup project. It used to be **ApsimNG**. Cheers.

**Comment by rcichota** [comment]
The need to compile Apsim.Shared separately after any major update can be very frustrating to many users that don't use Apsim regularly or are not programming savvy, I've get 'regular' complaints here. 
Now, much less often, but equally or even more frustrating is the need to copy binary files from /DeploymentSupport/. This only needs to be done when a completely new checkout is made, and thus always get people off-guard as it quite rare. we had a couple of people which got a new computer and took quite a while (and frustration) to figure what was going on and what was the solution. As far as I know the instructions are only buried in a closed issue (#2199). I know this may be a separate issue, but it is related the one here, especially in that it blocks compiling and creates unnecessary frustration. Maybe @hol430 or @zur003 have some ideas for getting this sorted? Or make this instructions easy to find...

**Comment by zur003** [comment]
My view is that the best approach would be to improve the documentation, with easy-to-find instructions. If we were building only on Windows, then I'd suggest a different approach, especially regarding the binaries in DeploymentSupport, but I think we want to continue to allow builds and development to be possible on OS X and Linux. Keeping platform-specific binaries isolated from cross-platform assemblies is important when trying to build on multiple platforms.

**Comment by hol430** [comment]
It might be possible to include a pre-build step in ApsimNG.csproj which copies the appropriate assemblies over if running on windows. I think I may have tried this before but didn't get it to work. I'll give this another look.

Regarding APSIM.Shared, I discussed this with @hol353 and we decided to move APSIM.Shared into the ApsimX repo. I'm working on this now. This would save the hassle of keeping both repos up to date which seems to be a problem for people who don't often use git. It also removes the need to manually restore nuget packages for APSIM.Shared. It also means that changes to APSIM.Shared will have to run through the test suite on jenkins.

The disadvantage is that a few of our repos (performance tests, build website, etc) use these utilities, and bugfixes in ApsimX won't propagate to the other services.
